src/state.rs

```diff
@@ -25,8 +23,31 @@ pub enum BadBackendError {
     UnknownQueue(String),
 }
 
+// This RAII wrapper streams a request body into memory so we can
+// examine it; when the wrapper is dropped, we stuff the body back
+// into the request so it can be forwarded to the backend.
+struct PeekBody<'a> {
+    bytes: Bytes,
+    body: &'a mut Body,
+}
+
+impl PeekBody<'_> {
+    async fn new(body: &mut Body) -> Result<PeekBody, hyper::Error> {
+        Ok(PeekBody {
+            bytes: hyper::body::to_bytes(&mut *body).await?,
```

Comment:
This commit was an interesting little borrow checker puzzle to solve. This reborrow in particular seems like it should be unneeded but the compiler requires it for some reason.

Maybe this `PeekBody` struct is worth moving to some kind of `util` module?